### ISSUE TITLE
Versioned Storages: use cases for Git operations

### ISSUE THREAD

**Issue by NShaforostov** [epam/cloud-pipeline#1807]
**Background**
Once the [Versioned Storage](https://github.com/epam/cloud-pipeline/issues/1805) (further - VS) support is implemented, users should have the ability to launch runs using that new type of storages.
Users should have the access to the VS during the run - similarly as to the folder of any mounted storage currently. Users should have the ability to read and write data (according to the permissions) to the VS.

**Approach**

Main details see in #1833. Here the main use cases are described.

Since the VS is a GitLab repository under the hood, there are some features of its using during the run:
- the VS should be cloned into the running instance (into the pre-configured location that is set by environment variable) - there can be the following ways:
  - launch a run from the VS page. In this case, the actual VS content will be automatically cloned into the certain path of the instance during the run initialization
  - cloning the VS content into the running instance. In this case, the actual VS content will be cloned into certain path of the previously launched instance
- to "save" all changes to the VS from the cloned directory after the computation, users shall manually perform the commit operation
- there shall be the ability to view diffs between the recent revision of the VS (from which the VS was cloned into the running instance) and "current" version of the cloned directory - to understand what will be committed
- there shall be the mechanism of the merge conflict resolving implemented - for cases when several users try to commit into the same VS with changes of the same files

Example scenario:

1. User launches the run from the VS page:
   - run starts with the VS cloned into some instance folder
2. User performs calculations over the cloned data (using e.g. _Remote Desktop_ or _SSH_)
3. User commits all changes to save them:
   - in the VS, a new revision is created

**Comment by NShaforostov** [comment]
### Main use cases. Using Versioned Storages in runs (Git operations)

- [Additional controls for VS usage during runs](#additional-controls)
- [Launch a run from the VS page](#launch)
- [Clone the VS content into the running instance](#clone-into-run)
- [Diffs for a cloned storage](#diffs)
- [Save content from the running instance to the VS](#commit-vs)
- [Change the revision in the running instance](#refresh-vs)

---

<h3 id="additional-controls">Additional controls for VS usage during runs</h3>

At the main **Dashboard** (_Active runs_ panel, run tiles), new item shall be added - **VCS** (_Version Control Storages_), e.g.:
![image](https://user-images.githubusercontent.com/45459424/113186069-dd765400-925f-11eb-87ab-47afd5e93ae0.png)
This item should appear only after full run initialization.

**VCS** item should contain sub-items:
- **Clone**
- **Save**
- **Refresh**
- **Diff**

Also, the similar control should appear at the **Run logs** page.

---

<h3 id="launch">Launch a run from the VS page</h3>

To launch a run from the VS page:
1. Open the VS that you wish to use during the run computation
2. Click the **Run** button in the right upper corner of the VS page
3. The launch form list shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/113180387-5aea9600-9259-11eb-8f96-669f4c9b2ddb.png)
   This form should contain:
   - sections where user can configure all necessary run settings (similar to general Launch form for the tool/pipeline). All main required fields ("Docker image", "Node type", "Disk", "Cloud Region") should be empty, by default
   - the cloning VS name in the form header
   - the dropdown list where user can select the revision of cloning VS
4. Select docker image (tool) and its version that you wish to run. Other mandatory setting values ("Node type", "Disk", "Cloud Region") shall automatically be received from the corresponding tool/tool version's settings or set manually.
5. Click the **Launch** button to launch the selected tool version with configured settings
6. Confirm the launch
7. Selected tool with configured settings shall be launched and the selected revision content of the VS shall be cloned into the running instance (during the run initialization). Example of the cloned directory in the instance (path is shown conditionally):
  ![image](https://user-images.githubusercontent.com/45459424/109180970-7ebf4580-779c-11eb-953f-c8a1b60e9864.png)

> **_Note_**:  that you can clone any non-latest revision of the VS to the run and then work with it from the run environment. But you will not have the ability to save changes in such case - all changes made in VS during the run will be lost after the run stop. Saving is able only when the **_latest_** version is being cloned to the run (default behavior).
> Therefore, when trying to run a tool, cloning non-latest VS revision, the corresponding warning should appear.

---

<h3 id="clone-into-run">Clone the VS content into the running instance</h3>

To clone the VS content into a running environment:
1. Hover the tile of the desired run at the _Active runs_ panel of the main **Dashboard** or open the **Run logs** page of this run
2. Click the **VCS** button
3. Select the **Clone** item in the list
4. The pop-up with VS list shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/111611034-11c42c00-87ed-11eb-9817-44869b27a023.png)
   This pop-up:
   - should contain only VS available for the current user 
   - for each VS there should be the ability to select the revision that will be cloned (_by default, the most recent revision (**latest**) is selected_)
   - shouldn't support multi-select
5. Select the VS and its revision to clone.
   Please note, that you can clone any non-latest revision of the VS to the run and then work with it from the run environment.
   But you will not have the ability to "save" (commit) changes in such case.
   "Saving" is able only when the **latest** version was cloned to the run.
   Therefore, when trying to clone non-latest revision, the warning should appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/111613539-acbe0580-87ef-11eb-9851-164867a3c5ec.png)
6. Click the **Clone** button to confirm
7. The progress bar for the operation performing shall appear
8. Once the operation is performed, the content of the VS of selected revision will be cloned into the pre-configured location of the running instance - similarly as described above, e.g.:
  ![image](https://user-images.githubusercontent.com/45459424/109180970-7ebf4580-779c-11eb-953f-c8a1b60e9864.png)

**_Note_**: user can clone any count of the available VS into the running instance

If user will try to clone into the instance the VS that has been cloned early - the warning shall appear with possible options:
- remove previously cloned content of this VS (clear the pre-configured location) from the instance
- cancel the cloning operation

---

<h3 id="diffs">Diffs for a cloned storage</h3>

At any point, user can request the diffs for a selected VS that was cloned into the running instance.

To view the diffs:
1. Hover the tile of the desired run at the _Active runs_ panel of the main **Dashboard** or open the **Run logs** page of this run
2. Click the **VCS** button
3. Select the **Diff** item in the list
4. The pop-up with VS list that were cloned to the instance shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/109995222-28fc1780-7d1f-11eb-8c86-8a82ecc5ebf6.png)
   This pop-up:
   - should contain only VS cloned by the current user into the current running instance
   - for each VS there the revision should be specified from which it was cloned
   - shouldn't support multi-select
5. Select the VS
6. Click the **View diffs** button to confirm
7. The pop up with the diffs between cloned VS revision and the current state of the VS content in the running instance shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/113279599-fa586900-92eb-11eb-8de9-577f2ad1fe0f.png)
   This pop-up shall contain:
   - two columns:
      - for the _origin_ state of the VS (content from VS revision that was cloned to the instance)
      - for the _changed_ state of the VS (current state of the VS content in the instance)
   - list of only changed files
   - for each changed _text_ file:
      - at least, the changed rows shall be shown
      - in the _origin_ column, edited/removed rows with numbers shall be shown in red color
      - in the _origin_ column, summary count of edited/removed rows shall be shown as `- <count> Removals`
      - in the _changed_ column, edited/added rows with numbers shall be shown in green color
      - in the _changed_ column, summary count of edited/added rows shall be shown as `+ <count> Additions`
   - for each changed _binary_ file, the changing of the file size shall be shown:
      - for the added file, in green color
      - for the deleted file, in red color
      - for the changed file, in blue color

---

<h3 id="commit-vs">Save content from the running instance to the VS</h3>

At any point, user can "save" changes made in the cloned content of the VS to the origin VS.

To perform the "save" operation from the running instance with the cloned VS:
1. Hover the tile of the desired run at the _Active runs_ panel of the main **Dashboard** or open the **Run logs** page of this run
2. Click the **VCS** button
3. Select the **Save** item in the list
4. The pop-up with VS list that were cloned to the instance shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/113279068-5cfd3500-92eb-11eb-9ffb-cc0170e624f2.png)
   This pop-up:
   - should contain only VS cloned by the current user into the current running instance
   - shouldn't support multi-select
5. Select the VS.
   Please note, that you can save changes only for the VS which content was cloned to the run from the latest VS revision (at the cloning moment).
   Therefore, near the VS based on the non-latest revision, the warning is shown. If this VS is selected, the **Proceed** button becomes disabled:
   ![image](https://user-images.githubusercontent.com/45459424/113279401-b2d1dd00-92eb-11eb-8726-4e95a7e54023.png)
6. Click the **Proceed** button to continue
7. The pop up to specify the commit message shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/110095215-ba19cf80-7dad-11eb-9d2f-dcec2b58a828.png)
8. Specify the commit message and click the **OK** button
9. The system tries to perform corresponding Git actions to remote VS.
10. If no conflict is detected, the pop up with the new revision confirmation shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/110098053-d1a68780-7db0-11eb-9e46-26dcb1ba3c01.png)
11. If any conflict is detected (e.g. any other user during the current run with the cloned VS changed the same file(s) as in the current run and "saved" these changes), the pop up with conflicting files list shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/113277822-ced47f00-92e9-11eb-9628-77df30b418ed.png)
   This pop-up shall contain:
     - list of only conflicting files
     - button to accept the current version of conflicting files as a conflict resolve (all conflicts simultaneously to the _changed_ state) - e.g., "**>> Accept current changes for all conflicting files**" at the picture above
     - button to accept the most recent VS repository version of conflicting files as a conflict resolve (all conflicts simultaneously to the _origin_ state) - e.g., "**<< Accept recent repository version for all conflicting files**" at the picture above
12. User selects conflicting file to resolve its conflicts by the corresponding button near the file name - e.g., "**Resolve conflict**" at the picture above. A new pop-up appears, e.g.:
  ![image](https://user-images.githubusercontent.com/45459424/116726849-81345a80-a9ec-11eb-9451-559361236706.png)
  This pop-up shall contain:
     - three columns:
        - for the _changed_ state of the conflicting file (current state of the file in the instance) - left column
        - for the _origin_ state of that conflicting file (current most recent revision of the file) - right column
        - for the _merging_ state (result) of the conflict resolve - center column
     - highlighted conflicting changes. These changes should be highlighted in all 3 columns. In picture above, they are highlighted in peach color
     - buttons to accept one of the full versions of conflicting file as a conflict resolve (all conflicts of the file simultaneously to one of the state - _changed_ or _origin_) - "**Accept this version**" buttons in left and right columns
     - for each conflict in the file (only for _text_ files):
        - ability to "get" separately each of the conflicting changes (from _changed_ file version or _origin_) as resolving of the conflict - by the corresponding buttons ![image](https://user-images.githubusercontent.com/45459424/112998968-33bb9800-9177-11eb-9466-40cfe4faf861.png) and ![image](https://user-images.githubusercontent.com/45459424/112999041-433ae100-9177-11eb-9ec9-ad5e04f0423c.png) near the conflict
        - ability to edit the conflicting file manually - by the corresponding "Edit" icon near the file name. **_TBD_**: in this case, the Cloud Pipeline's text file editor shall be open, that shall display "merged" file with conflicting parts in the following view:
     ![image](https://user-images.githubusercontent.com/45459424/110105849-f3583c80-7db9-11eb-8d82-4d56cb1aea23.png)
     - if conflicting file is _binary_:
        - ability to "get" one of the full file version (_origin_ or _changed_) as resolving of the conflict - by the corresponding button **Accept this version** (for each version, the file size is shown)
     - also for _text_ conflicting files, non-conflicting merged changes should be displayed. These changes are being merged automatically, user can see only their merge result:
        - if in any version (_changed_ or _origin_), there a text addition/modification was - it should be highlighted in the corresponding version and in the result column. In picture above, it is highlighted in blue color
        - if in any version (_changed_ or _origin_), there a text deletion was - it should be highlighted in that version, where this text is still in place. In other version and in result column, only the place, where this text was before deletion, should be marked by the line - in picture above, see an example, it is highlighted in gray color.
13. All conflicts should be resolved. If all conflicts of the specific file are resolved, the button "**Accept result**" becomes available.
  Example of the conflicts resolve by the button "**Accept this version**" in the _origin_ column:
  ![image](https://user-images.githubusercontent.com/45459424/116727045-ba6cca80-a9ec-11eb-942f-df4918dec2f8.png)
  To cancel the conflict resolving use the corresponding cross-button near the resolve - ![image](https://user-images.githubusercontent.com/45459424/113024085-0549b700-918f-11eb-9a3b-647ebf24e0a6.png).
  Example of the manual resolve of conflicts:
  ![image](https://user-images.githubusercontent.com/45459424/116727358-1d5e6180-a9ed-11eb-814b-21e6ca1659c8.png)
  Click the button "**Accept result**".
  > **_Note_**: if user wants to postpone for a while conflicts resolving, he(she) can click the **Cancel** button or the cross-button at the right upper corner of the pop-up. In this case, the VS will remain in the _conflicted_ state (will wait for the conflict resolving).
  > On the other hand, user can click the **Abort** button. This should cancel all performed but unsaved resolves and return the VS to the "clean" _changed_ state - as it was before the saving attempt. After that, user can retry the saving operation from the beginning.
14. The pop up with conflicting files list shall appear again, e.g.:
  ![image](https://user-images.githubusercontent.com/45459424/113277972-017e7780-92ea-11eb-9da8-f929d49133a8.png)
  All conflicts should be resolved. If conflicts in all conflicting files are resolved, the button "**Apply and save**" becomes available, e.g.:
  ![image](https://user-images.githubusercontent.com/45459424/113278182-44404f80-92ea-11eb-90f0-98a5dde6825a.png)
  **_Note_**: user can re-resolve any conflict before clicking the button "**Apply and save**", using ways described above
  Click the button "**Apply and save**".
  The pop up similar to the one from step 10 appears.

---

<h3 id="refresh-vs">Change the revision (and content) of the VS in the running instance</h3>

At any point, user can "refresh" the VS content in the running instance - change the VS revision, from which the content should be cloned. By default, this feature should be used for update the content to the latest revision (most recent) - as VS _origin_ content could be changed by other users during the run, e.g. But also, this feature should allow user to change the revision to any non-latest - just to "view" the state of the VS in certain revision.

Different variants of the VS content "refresh" have the following restrictions:
- if user "refreshes" the VS content to any **NON-LATEST** revision - all performed current changes of the cloned VS content will be lost (no matter from which revision the current content was cloned). **_Note_**: _user should receive warning message in such case_
- if user "refreshes" the VS content to **LATEST** revision but the current VS revision was **NON-LATEST** at its cloning moment - all performed current changes of the cloned VS content will be lost. **_Note_**: _user should receive warning message in such case_
- if user "refreshes" the VS content to **LATEST** revision and the current VS revision was **LATEST** at its cloning moment - all performed current changes should be merged during the "refresh" (see details below)

To perform the "refresh" operation from the running instance with the cloned VS:
1. Hover the tile of the desired run at the _Active runs_ panel of the main **Dashboard** or open the **Run logs** page of this run
2. Click the **VCS** button
3. Select the **Refresh** item in the list
4. The pop-up with VS list that were cloned to the instance shall appear, e.g.: 
   ![image](https://user-images.githubusercontent.com/45459424/113268236-d727bc80-92df-11eb-8cb9-aab6319702ec.png)
   This pop-up:
     - should contain only VS cloned by the current user into the current running instance
     - for each VS there the revision should be specified from which it was cloned or the revision to which the VS was "refreshed"
     - shouldn't support multi-select
5. Select the VS. For the selected VS, the dropdown with the available revisions list should appear (by default - "latest" revision for any VS). The **Refresh** button becomes enabled, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/113270470-2b33a080-92e2-11eb-84f7-cdb48101d3ba.png)
   Select the revision in the dropdown list.
   Please note, that trying to refresh to any **NON-LATEST** revision or from base **NON-LATEST** revision should be accompanied with the corresponding messages as described above, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/113275780-d2670680-92e7-11eb-928e-135df03ad5a0.png)
6. Click the **Refresh** button to continue
7. The system tries to perform corresponding Git actions from the remote VS.
8. If no conflict is detected, the pop up with the commit confirmation shall appear, e.g.:
   ![image](https://user-images.githubusercontent.com/45459424/110618772-1e6ed180-81a8-11eb-845d-2118e596beb7.png)
9. In case, when the "refresh" is being performed to **LATEST** revision and the current VS revision was **LATEST** at its cloning moment - conflicts may appear (e.g. any other user during the current run with the cloned VS changed the same file(s) as in the current run and saved these changes). In such case, the pop up with conflicting files diffs shall appear. Mechanism of the conflicts resolve should be the similar as described [above](#commit-vs).

**Comment by maryvictol** [comment]
Tags for differences are different for **Commit changes** pop-up from the running instance with the cloned VS and for VS Revision History:
**CREATED** for **Commit changes** pop-up <-> **ADDED** for VS Revision History
![image](https://user-images.githubusercontent.com/47791878/119497816-a7d87d80-bd6d-11eb-817b-72e0883fe62f.png)
![image](https://user-images.githubusercontent.com/47791878/119497773-9d1de880-bd6d-11eb-9d11-ea70fbeb6ce5.png)

**MODIFIED** for **Commit changes** pop-up <-> **CHANGED** for VS Revision History

**Comment by maryvictol** [comment]
File with spaces in the name created in cloned VS content gives error `Unexpected response status code: 404 Message: {"message":"404 File Not Found"}` at attemption rename it in the VS or see it's preview.
**Steps to reproduce:**
1. Clone the VS content into the running instance
2. Create file with spaces in the name and commit it
3. Open VS and click created file name to open its preview => error message is shown on Information panel
4. Click **Edit** button next created file name. Change name. Click OK button => error message appears

**Comment by maryvictol** [comment]
Folder created in cloned VS content can't be renamed in VS.
**Steps to reproduce:**
1. Clone the VS content into the running instance
2. Create folder with file and commit it
3. Open VS and try to rename created folder

**Actual result:**
Error message `Error: folder <folder name> not found.` appears

**Comment by maryvictol** [comment]
@rodichenko Launch a run from the VS page as non-admin user gives error `Unexpected response status code: 404 Message: {"message":"404 File Not Found"}`.
**Steps to reproduse:**
1. Login as non-admin user
2. Open any VS 
3. Click **Run** button
4. Select Tool and click **Launch** button (or use Run custom button)

**Comment by rodichenko** [comment]
> @rodichenko Launch a run from the VS page as non-admin user gives error `Unexpected response status code: 404 Message: {"message":"404 File Not Found"}`.
> **Steps to reproduse:**
> 
> 1. Login as non-admin user
> 2. Open any VS
> 3. Click **Run** button
> 4. Select Tool and click **Launch** button (or use Run custom button)

@mzueva @ekazachkova could you check this please?

**Comment by maryvictol** [comment]
The preview and Revision history remain on the Information panel after deleting the file.
**Step to reproduce:**
1. Open VS
2. Click on file name to open Information panel
3. Click Delete icon near the file
![image](https://user-images.githubusercontent.com/47791878/120554845-ec65b800-c402-11eb-8b57-32d2f3e3d5fc.png)

**Comment by maryvictol** [comment]
@rodichenko Incorrect tooltip `[object Object]` is shown for revision field on **Select Version Storage** pop-up.
![image](https://user-images.githubusercontent.com/47791878/120625288-37ff7c80-c46a-11eb-9756-84aaacd8ad31.png)

**Comment by rodichenko** [comment]
> @rodichenko Incorrect tooltip `[object Object]` is shown for revision field on **Select Version Storage** pop-up.
> ![image](https://user-images.githubusercontent.com/47791878/120625288-37ff7c80-c46a-11eb-9756-84aaacd8ad31.png)

@maryvictol 74c60478a83e0ba729aff593000270c725990c3d fixes that

**Comment by ekazachkova** [comment]
> @rodichenko Launch a run from the VS page as non-admin user gives error `Unexpected response status code: 404 Message: {"message":"404 File Not Found"}`.
> **Steps to reproduse:**
> 
> 1. Login as non-admin user
> 2. Open any VS
> 3. Click **Run** button
> 4. Select Tool and click **Launch** button (or use Run custom button)

@maryvictol  fixed

**Comment by maryvictol** [comment]
@rodichenko 
At case of Launch a run from the VS page message `current revision is not the latest` is shown in the **Versioned Storage** menu for ***Save*** item.
![image](https://user-images.githubusercontent.com/47791878/120983614-8a9aaa80-c782-11eb-86cf-db34ee96b04a.png)

**Comment by rodichenko** [comment]
> @rodichenko
> At case of Launch a run from the VS page message `current revision is not the latest` is shown in the **Versioned Storage** menu for _**Save**_ item.
> ![image](https://user-images.githubusercontent.com/47791878/120983614-8a9aaa80-c782-11eb-86cf-db34ee96b04a.png)

@ekazachkova (@mzueva) could you please take a look? We've got `detached` == `true` for the repo after running VS via POST `/run`:

**After launching**:
```sh
> git log
commit 65867d3e2ae4aa411bc3d9e17d1880216fde05f2 (HEAD, origin/master, origin/HEAD, master)

> git status
HEAD detached at 65867d3
nothing to commit, working tree clean
```

**After refresh operation**:
```sh
> git log
commit 65867d3e2ae4aa411bc3d9e17d1880216fde05f2 (HEAD -> master, origin/master, origin/HEAD)

> git status
On branch master
Your branch is up to date with 'origin/master'.
nothing to commit, working tree clean
```

**Comment by maryvictol** [comment]
@mzueva, @ekazachkova 
Error message `'NoneType' object has no attribute '_pointer'` is shown when user try to change the VS revision, from which the content should be cloned, to morw recent without Refresh.
![image](https://user-images.githubusercontent.com/47791878/124460990-c3856980-dd98-11eb-8a25-725acb9fd447.png)

**Steps to reproduce:**
1.  Launch the run from the VS page or Clone VS to any running instance
2. Open VS. 
3. Make any changes (for ex. create new file or folder)
4. Return to the run from step 1
5. Click _Versioned Storages_ link >> Select _Checkout revision_ option
6. Select the most recent revision from the drop-down list >> Click _Change Revision_ button

**Comment by maryvictol** [comment]
@SilinPavel 
Delete file commit disapppears from history at case of file is deleted and updated by 2 users at the same time.

**Steps to reproduce:**
1. Open VS and create file1
2. ***User1:*** Launch run. Clone VS to running instance. 
3. Edit file1 using ssh.
4. ***User2:*** Open VS and delete file1
5. ***User1:***  Open logs page of run from step 2.
6. Click **Versioned Storage** link. Select *Save* option from menu 
7. Commit changes.

**Actual results:**
1. Open **Checkout revision** pop up: revision id that is shown as current version doesn't exist in the list of revisions
2. Open *Revision history* on the VS page: User2's commit about delete file1 isn't shown in the History. Last revision Id doesn't correspond current version VS.

**Comment by SilinPavel** [comment]
@maryvictol should be fixed by b9fc054 and 01ef32e

**Comment by maryvictol** [comment]
> @maryvictol should be fixed by [b9fc054](https://github.com/epam/cloud-pipeline/commit/b9fc054ad98261aecf1c96cb3e73494a113f77c4) and [01ef32e](https://github.com/epam/cloud-pipeline/commit/01ef32edd533f9a1196741836a58bd4417f4aabc)

Verified as fixed

**Comment by NShaforostov** [comment]
Docs were added via #3217 and located [here](https://github.com/epam/cloud-pipeline/blob/develop/docs/md/manual/08_Manage_Data_Storage/8.13._Versioned_storages.md#git-operations-with-versioned-storages).
